packages/react/src/components/TreeView/TreeView.js

```diff
@@ -0,0 +1,196 @@
+  size = 'default',
+  ...rest
+}) {
+  const { current: treeId } = useRef(rest.id || uniqueId());
+  const treeClasses = classNames(className, `${prefix}--tree`, {
+    [`${prefix}--tree--${size}`]: size !== 'default',
+  });
+  const treeRootRef = useRef(null);
+  const treeWalker = useRef(treeRootRef?.current);
+  const [selected, setSelected] = useState(preselected);
+  const [active, setActive] = useState(prespecifiedActive);
+  function handleTreeSelect(event, node = {}) {
+    const { id: nodeId } = node;
+    if (multiselect && (event.metaKey || event.ctrlKey)) {
+      if (!selected.includes(nodeId)) {
+        setSelected(selected.concat(nodeId));
+      } else {
+        setSelected(selected.filter((selectedId) => selectedId !== nodeId));
+      }
+    } else {
+      setSelected([nodeId]);
+      setActive(nodeId);
+    }
+    if (onSelect) {
+      onSelect(event, node);
+    }
+  }
+  let focusTarget = false;
+  const nodesWithProps = React.Children.map(children, (node) => {
+    const sharedNodeProps = {
+      active,
+      depth: 0,
+      onTreeSelect: handleTreeSelect,
+      selected,
+      tabIndex: (!node.props.disabled && -1) || null,
+    };
+    if (!focusTarget && !node.props.disabled) {
+      sharedNodeProps.tabIndex = 0;
+      focusTarget = true;
+    }
+    if (React.isValidElement(node)) {
+      return React.cloneElement(node, sharedNodeProps);
+    }
+  });
+
+  function handleKeyDown(event) {
+    event.stopPropagation();
+    if (matches(event, [keys.ArrowUp, keys.ArrowDown])) {
+      event.preventDefault();
+    }
+    treeWalker.current.currentNode = event.target;
+    let nextFocusNode;
+    if (match(event, keys.ArrowUp)) {
+      nextFocusNode = treeWalker.current.previousNode();
+    }
+    if (match(event, keys.ArrowDown)) {
+      nextFocusNode = treeWalker.current.nextNode();
+    }
+    if (nextFocusNode && nextFocusNode !== event.target) {
+      Array.prototype.forEach.call(
+        treeRootRef?.current?.querySelectorAll('[tabIndex="0"]') ?? [],
+        (item) => {
+          item.tabIndex = -1;
+        }
+      );
+      nextFocusNode.tabIndex = 0;
+      nextFocusNode.focus();
+    }
+    if (rest.onKeyDown) {
+      rest.onKeyDown(event);
+    }
+  }
+
+  useEffect(() => {
+    treeWalker.current =
+      treeWalker.current ??
+      document.createTreeWalker(treeRootRef?.current, NodeFilter.SHOW_ELEMENT, {
+        acceptNode: function (node) {
+          if (node.classList.contains(`${prefix}--tree-node--disabled`)) {
+            return NodeFilter.FILTER_REJECT;
+          }
+          if (node.matches(`li.${prefix}--tree-node`)) {
+            return NodeFilter.FILTER_ACCEPT;
+          }
+          return NodeFilter.FILTER_SKIP;
+        },
+      });
+  }, []);
+
+  useEffect(() => {
+    if (preselected.length) {
+      setSelected(preselected);
+    }
+    if (prespecifiedActive) {
+      setActive(prespecifiedActive);
+    }
+  }, [preselected, prespecifiedActive]);
```

Comment:
It seems like this would re-run if the parent component ever updated unless the values were memoized. Tried to recreate this over in storybook like:

```jsx
function ParentComponent() {
  const [, forceUpdate] = useState([]);
  const selected = ['5'];
  return (
    <>
      <button onClick={() => forceUpdate([])>Update</button>
      <TreeView {...props} selected={selected} />
    </>
  );
}
```

![demo](https://user-images.githubusercontent.com/3901764/89798291-9b8e7700-daf1-11ea-8381-ffed494f79b0.gif)

With `useMemo` on the value I think you get the intended outcome (update does not change selected) but we could try [this pattern](https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops) like we do in other parts of the codebase, or not support switching from controlled <-> uncontrolled since it's unclear what's right in this case.

Reply:
so if I understand correctly, we would perform an equality check on the selected/active node arrays before setting the state?